Added redirects for old React versions

bvaughn opened the pull request:
Use [Netlify's redirects API](https://www.netlify.com/docs/redirects/)- and the fact that Netlify caches immutable builds of older versions of the site- to expose previous versions of the React documentation via a [**new "React Versions" page**](https://deploy-preview-801--reactjs.netlify.com/versions).

I did the following to identify the correct revisions to use:
1. Lookd at the [`CHANGELOG`](https://github.com/facebook/react/blob/master/CHANGELOG.md) for release dates.
2. Cross-references [Netlify build history for the `master` branch](https://app.netlify.com/sites/reactjs/deploys?filter=master)
3. Picked the most recent build _before_ we updated the site (e.g. merged a PR that adds new API docs)

Unfortunately, our Netlify history begins with version 16 RC3 (so we can't link back to an old 15.6 release).

Since Netlify has built this PR preview as [5ad0def873f2cf6608e0d48a--reactjs.netlify.com](https://5ad0def873f2cf6608e0d48a--reactjs.netlify.com/), we should be able to test these new redirects as:
* 16.0: https://5ad0def873f2cf6608e0d48a--reactjs.netlify.com/version/16.0
* 16.1: https://5ad0def873f2cf6608e0d48a--reactjs.netlify.com/version/16.1
* 16.2: https://5ad0def873f2cf6608e0d48a--reactjs.netlify.com/version/16.2

A `robots.txt` file has also been added- along with `rel="nofollow"` attributes- to prevent search crawlers from indexing older versions of the site. (We want the latest release to be what shows up in search engines at all times.)

This versioning strategy should also work well by default with any [localization strategy](https://github.com/reactjs/reactjs.org/issues/82) we use.

**TODO**:
- [x] Add a UI mechanism for viewing old revisions
- [x] Don't let search engines crawl older versions
- [x] Make versions configurable via a single YAML file
- [ ] Determine if there's a way to use a subdomain of reactjs.org for versions rather than the netlify URL

Resolves issue #85

### Screenshot
<img width="402" alt="screen shot 2018-04-14 at 9 53 34 am" src="https://user-images.githubusercontent.com/29597/38770561-bee19cc0-3fc9-11e8-9b70-bc1b01773eec.png">

gaearon commented:
Will Netlify ever remove them?

consultwithmike commented:
This is awesome!

Looks good at first with this screenshot:

<img width="676" alt="screen shot 2018-04-13 at 11 15 12 am" src="https://user-images.githubusercontent.com/10493061/38748638-08144e88-3f0c-11e8-95c8-6e191c5f3ee0.png">

and remains good if I use the menus to navigate:

<img width="668" alt="image" src="https://user-images.githubusercontent.com/10493061/38748682-299425ec-3f0c-11e8-8c8d-b8e50a9f86b9.png">

but if I use Search it will use the latest version:

<img width="645" alt="image" src="https://user-images.githubusercontent.com/10493061/38748714-3f743df2-3f0c-11e8-9716-058962f8a4d1.png">

bvaughn commented:
> Will Netlify ever remove them?

I don't think so. On their [features page](https://www.netlify.com/features/) they say each URL is permanent:
![screen shot 2018-04-13 at 10 18 09 am](https://user-images.githubusercontent.com/29597/38748756-04c92c92-3f04-11e8-8547-6b8429391f03.png)

gaearon commented:
Well, as long as Netlify is in business 😛

alexkrolick commented:
Search always uses the newest version, even for deployed previews and localhost. Probably should be tackled separately.

consultwithmike commented:
@alexkrolick we can tackle it separately with a different issue if we want, and I fully suspected that there was a code change coming down the pipeline, I just wanted to drop that information on the issue so we can get it handled.

alexkrolick commented:
Issue for versioned search: https://github.com/reactjs/reactjs.org/issues/803

bvaughn commented:
I added a drop-down toggle button for version switching. I'd love to get feedback, but I think I'm going to immediately replace it with a "older versions" index page because:
* It would not work for mobile (since we don't even show the version number)
* This would not work well once we have more than a few versions

I just wanted to add it so it would be available in the history of this PR.

bvaughn commented:
Okay, I changed the version link in the header to link to a "Versions" page, which points to both the CHANGELOG _and_ docs for recent releases of the site.

alexkrolick commented:
Is there some way to source both the redirects and the page content from the same source? Like `versions.yml`:
```
versions:
	- title: 16.0
	  version: 16.0.0
	  url: 5ad0def873f2cf6608e0d48a--reactjs.netlify.com/version/16.0
	- title: 16.1
	  version: 16.1.1
	  url: 5ad0def873f2cf6608e0d48a--reactjs.netlify.com/version/16.1
```

bvaughn commented:
Yeah! Let me see if I can whip up another Gatsby startup plugin thingy for that.

bvaughn commented:
Ok. I added a Gatsby plugin that reads version values from a YAML file and adds them as Netlify redirect rules. Also updated the Versions page to read from this same YAML file. It's a good bit of extra code compared to the previous version- but it simplifies the process of adding new versions.

gaearon commented on src/pages/versions.js:
```diff
@@ -0,0 +1,49 @@
+          <ul>
+            {versions.map(version => (
+              <li key={version.title}>
+                <a href={version.path} rel="nofollow">
```
Since the top link used to always lead to a release, what do you think about making releases first-class entities on this page?

Something like

```
# React 16.2.0

* Changelog
* Documentation

# React 16.1.0

* Changelog
* Documentation

```

bvaughn replied:
I dig it.

bvaughn commented:
I think we should move forward with this and follow up with the vanity subdomain URLs (assuming Netlify supports this- I've reached out to them about it on Twitter). This is a useful feature for the community, even as-is (with ugly URLs). 😄

bvaughn commented:
cc @danoc in case you know of a way to do this already ^

bvaughn commented:
I wonder if a modified version of this plug-in could be useful for other Gatsby sites? Or maybe as an addition to the existing [`gatsby-plugin-netlify` plugin](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-netlify/)... cc @KyleAMathews

KyleAMathews commented:
Cool idea! The thought of duplicating documentation every time you do a new release has always filled me loathing (more-or-less) so this seems tons better. Relatively few people care about the older docs so pointing them to a snapshot seems like the least work for the most benefit.

It seems like this could be a lot simpler by fixing the redirect bug? Then you'd just need to read in the yaml file and create the redirects and the existing gatsby-plugin-netlify would setup the server redirects for you. I think all we'd need to do is to add a check [here](https://github.com/gatsbyjs/gatsby/blob/462e3b5b57c444e8d6acd3321422e29c06d69674/packages/gatsby/src/redux/actions.js#L967) if the to path is absolute and if so, don't add the pathPrefix.

danoc commented:
> cc @danoc in case you know of a way to do this already ^

AFAIK, the only way to do this right now is with the "branch subdomains" feature in Netlify. Would be great if there was a way to "tag" certain Netlify deploys and provide clean-lookin' URLs. 😄

bvaughn commented:
FYI there is a somewhat hacky option we could use to generate vanity subdomains (e.g. 16-0.reactjs.org) as per Netlify's tweet: https://twitter.com/NetlifySupport/status/985338587955810304

I'm not convinced this is worth it, TBH.

bvaughn commented:
> It seems like this could be a lot simpler by fixing the redirect bug? Then you'd just need to read in the yaml file and create the redirects and the existing gatsby-plugin-netlify would setup the server redirects for you.

Yes~ agreed! I assumed this bug was somewhat tricky to fix since it's been around for a while, but that would be an easier solution for sure. 😄

KyleAMathews commented:
Nah, more no one has cared yet :-)

bvaughn commented:
**Update on the vanity subdomains**: [According to Netlify support](https://twitter.com/NetlifySupport/status/985338587955810304), the only way to have a subdomain like 16-0.reactjs.org point to a cached snapshot of the site would be to create a Git branch from the commit that generated that build, and then setup a [branch subdomain](https://www.netlify.com/docs/custom-domains/#branch-subdomains). This approach would, I suppose, have the small added benefit of allowing us to update old branches if we ever needed to- but it also adds a bit of setup work each time we want to create a version.

I also tried another approach- using a [200 proxy redirect](https://www.netlify.com/docs/redirects/#proxying) to gives us URLs like reactjs.org/16.0. Unfortunately it didn't work. (The latest version of the site loaded, rather than the older one I was trying to proxy.)

gaearon commented:
Now I'm hungry 😢
